tests/unit/action/test_bonsai_asset.py

```diff
@@ -0,0 +1,243 @@
+    @pytest.mark.parametrize("name,args,required,typ", [
+        # Required values must match the selected type.
+        ("a", dict(a=3), True, int),
+        ("a", dict(a=3.3), True, float),
+        ("a", dict(a="b"), True, str),
+        ("a", dict(a=[]), True, list),
+        ("a", dict(a={}), True, dict),
+        # Optional values are not checked for type-correctness.
+        ("a", dict(), False, int),
+        ("a", dict(), False, float),
+        ("a", dict(), False, str),
+        ("a", dict(), False, list),
+        ("a", dict(), False, dict),
+    ])
+    def test_valid_values(self, name, args, required, typ):
+        bonsai_asset.validate(name, args, required, typ)
+
+    def test_missing_required(self):
+        with pytest.raises(errors.Error, match="required"):
+            bonsai_asset.validate("a", {}, True, str)
+
+    def test_invalid_type(self):
+        with pytest.raises(errors.Error, match="should"):
+            bonsai_asset.validate("a", dict(a=3), True, str)
+
+
+class TestValidateArguments:
+    def test_valid_minimal_args(self):
+        bonsai_asset.ActionModule.validate_arguments(dict(
+            name="abc", version="1.2.3",
+        ))
+
+    def test_valid_all_args(self):
+        bonsai_asset.ActionModule.validate_arguments(dict(
+            name="abc", version="1.2.3", rename="def",
+            labels={}, annotations={},
+        ))
+
+    def test_invalid_name(self):
+        with pytest.raises(errors.Error, match="name"):
+            bonsai_asset.ActionModule.validate_arguments(dict(
+                name=1.234, version="1.2.3",
+            ))
+
+    def test_missing_name(self):
+        with pytest.raises(errors.Error, match="name"):
+            bonsai_asset.ActionModule.validate_arguments(dict(
+                version="1.2.3",
+            ))
+
+    def test_invalid_version(self):
+        with pytest.raises(errors.Error, match="version"):
+            bonsai_asset.ActionModule.validate_arguments(dict(
+                name="abc", version=1.2,
+            ))
+
+    def test_missing_version(self):
+        with pytest.raises(errors.Error, match="version"):
+            bonsai_asset.ActionModule.validate_arguments(dict(
+                name="abc",
+            ))
+
+    def test_invalid_rename(self):
+        with pytest.raises(errors.Error, match="rename"):
+            bonsai_asset.ActionModule.validate_arguments(dict(
+                name="abc", version="1.2.3", rename=1,
+            ))
+
+    def test_invalid_labels(self):
+        with pytest.raises(errors.Error, match="labels"):
+            bonsai_asset.ActionModule.validate_arguments(dict(
+                name="abc", version="1.2.3", labels=1,
+            ))
+
+    def test_invalid_annotations(self):
+        with pytest.raises(errors.Error, match="annotations"):
+            bonsai_asset.ActionModule.validate_arguments(dict(
+                name="abc", version="1.2.3", annotations=1,
+            ))
+
+
+class TestBuildAssetArgs:
```

Comment:
Can we add a test to check that `auth` is in the result if provided as argument?